src/sml.cpp

```diff
@@ -289,28 +310,31 @@ void pow(double &val, signed char &scaler)
 void smlOBISByUnit(long long int &val, signed char &scaler, sml_units_t unit)
 {
   unsigned char i = 0, pos = 0, size = 0, y = 0;
+  sml_states_t type;
   val = -1; /* unknown or error */
   while (i < listPos) {
     pos++;
-    if (pos == 4 && listBuffer[i + 1] != unit) {
-      /* if unit at position 4 is not 0x1e (wh) return unknown */
+    size = (int)listBuffer[i++];
+    type = (sml_states_t)listBuffer[i++];
+    if (pos == 4 && listBuffer[i] != unit) {
+      /* return unknown (-1) if unit does not match */
       return;
     }
     if (pos == 5) {
-      scaler = listBuffer[i + 1];
+      scaler = listBuffer[i];
     }
     if (pos == 6) {
-      size = (int)listBuffer[i];
       val = 0;
+      y = size;
+      if (type == SML_DATA_SIGNED_INT && (listBuffer[i] & (1 << 7))) {
+        // negative value
+        val = 0xFFFFFFFFFFFFFFFF;
+      }
       for (y = 0; y < size; y++) {
-        // convert value
-        // SML_LOG("size = %i, pos = %i, bit = %i\n", size, y + 1,
-        //        (8 * (size - 1)) - (8 * y));
-        val |= (long long int)listBuffer[i + y + 1]
-               << ((8 * (size - 1)) - (8 * y));
+        val = (val << 8) | listBuffer[i + y];
```

Comment:
This line has me confused. For a negative number, you would OR the initial 0xFF with whatever byte way received from SML - this should always result in 0xFF. I would expect to see something along these lines:

```suggestion
        if (type == SML_DATA_SIGNED_INT && (listBuffer[i] & (1 << 7))) {
          val = (val << 8) & listBuffer[i + y];
        } else {
          val = (val << 8) | listBuffer[i + y];
        }
```

Did the original code pass the test?

As an additional suggestion: Maybe factor out the entire check into a `const bool` near lin 329 to avoid repeating the check code so many times.